src/main/java/org/openrefine/extensions/commons/utils/CommonsImportingController.java

```diff
@@ -206,60 +207,67 @@ static public void parse(
         /* get user-input from the Post request parameters */
-        String cmtitle = JSONUtilities.getString(options, "categoryInput", null);
-        if (!cmtitle.startsWith("Category:")) {
-            cmtitle = "Category:" + cmtitle;
+        JsonNode categoryInput = options.get("categoryJsonValue");
+        List<String> categories = new ArrayList<>();
+        for (JsonNode category: categoryInput) {
+            categories.add(category.get("category").asText());
         }
-        String pageName = cmtitle;
-        String fileSource = pageName;//FIXME: add filename
         String apiUrl = "https://commons.wikimedia.org/w/api.php";
 
-        setProgress(job, fileSource, 0);
+        setProgress(job, categories.get(0), 0);// handles the first category
 
         TabularImportingParserBase.readTable(
                 project,
                 job,
-                new FilesBatchRowReader(job, fileSource, cmtitle, apiUrl),
+                // FIXME: FilesBatchRowReader will handle progress of the other categories
+                new FilesBatchRowReader(job, categories, apiUrl),
                 limit,
                 options,
                 exceptions
         );
-        setProgress(job, fileSource, 100);
+        setProgress(job, categories.get(categories.size()-1), 100);
     }
 
-        static private void setProgress(ImportingJob job, String fileSource, int percent) {
-            job.setProgress(percent, "Reading " + fileSource);
+        static private void setProgress(ImportingJob job, String category, int percent) {
+            job.setProgress(percent, "Reading " + category);
         }
 
         static protected class FilesBatchRowReader implements TableDataReader {
             final ImportingJob job;
-            final String fileSource;
             String apiUrl;
             HttpUrl urlBase;
             HttpUrl urlContinue;
             JsonNode files;
-            String cmtitle;
+            List<String> categories;
+            String category;
             String cmcontinue;
             private int indexRow = 0;
+            private int indexCategories = 0;
             List<Object> rowsOfCells;
 
-            public FilesBatchRowReader(ImportingJob job, String fileSource, String cmtitle, String apiUrl) throws IOException {
+            public FilesBatchRowReader(ImportingJob job, List<String> categories, String apiUrl) throws IOException {
 
                 this.job = job;
-                this.fileSource = fileSource;
-                this.cmtitle = cmtitle;
+                this.categories = categories;
                 this.apiUrl = apiUrl;
-                getFiles();
+                getCategory();
 
             }
+            public void getCategory() throws IOException {
+                // FIXME
+                for (int i=0; i < categories.size(); i++) {
+                    System.out.println(categories.get(i));
+                    getFiles(categories.get(i));
+                }
```

Comment:
Here, you are calling `getFiles` for all categories in order. What `getFiles` does is fetch the first page of filenames for each category and store it in the `files` attribute. So, with this loop, you are overriding the `files` attributes many times, until the last category.

So, after the initialization of `FilesBatchRowReader`, the `files` attribute will only contain files from the last category. That means the first categories will not make it to the project at all.